src/build_context.rs

```diff
@@ -1,133 +1,104 @@
-use pyo3::intern;
+use std::collections::hash_map::Entry;
+
 use pyo3::prelude::*;
-use pyo3::types::{PyDict, PyList};
 
 use ahash::{AHashMap, AHashSet};
 
-use crate::build_tools::{py_err, py_error_type, SchemaDict};
+use crate::build_tools::py_err;
 use crate::serializers::CombinedSerializer;
 use crate::validators::{CombinedValidator, Validator};
 
-#[derive(Clone, Debug)]
-struct Slot<T> {
-    slot_ref: String,
-    op_val_ser: Option<T>,
-}
-
-pub enum ThingOrId<T> {
-    Thing(T),
-    Id(usize),
-}
-
 /// `BuildContext` is used to store extra information while building validators and type_serializers
 #[derive(Clone, Debug)]
 pub struct BuildContext<T> {
-    /// set of used refs, useful to see if a `ref` is actually used elsewhere in the schema
-    used_refs: AHashSet<String>,
-    /// holds validators/type_serializers which reference themselves and therefore can't be cloned and owned
-    /// in one or multiple places.
-    slots: Vec<Slot<T>>,
-    /// holds validators/type_serializers which need to be accessed from multiple other validators/type_serializers
-    /// and therefore can't be owned by them directly.
-    reusable: AHashMap<String, T>,
+    // Mapping of reference strings to indexes in the definitions Vec
+    // As we recurse through the input when we find:
+    // - A definitions-reference (`{'type': 'definitions-reference', 'schema_ref': '...'}`)
+    //   then we store that `schema_ref` -> `definitions.len()`
+    //   so that the value is the index into definitions.
+    //   We also set a None value in definitions.
+    //   Then we create a DefinitionsReferenceValidator that stores the index (value) from this mapping.
+    //   At runtime if it needs to call into the validator it simply looks it up by index.
+    // - Any schema with a `'ref'`: before recursing we store the ref here and set a None
+    //   value in definitions.
+    //   As we recurse if a definitions-reference to this
+    //   Then after recursing and completing the CombinedValidator we replace that None value with the
+    //   newly built validator. Now anything that
+    refs: AHashMap<String, usize>,
+
+    // Here we hold an *optional* reference to a definition validator
+    // Any time we encounter a schema with a `'ref'` we build a DefinitionsReferenceValidator that points
+    // to an index in this Vector
+    // At the end of building we check if any values are None, which indicates an unsatisfied reference
+    definitions: Vec<Option<T>>,
 }
```

Comment:
It isn't too hard to add a basic ref-count system here so we know which references have more than one referrer and hence which can be inlined or not. But I got bogged down in implementation multiple times.